File: store/cxxsp/mysql.cpp
#include "mysql.h"
#include <stdexcept>

using namespace std;

Mysql mysql;

Mysql::Mysql() {
    handle = mysql_init(NULL);
    handle = mysql_real_connect(handle, "localhost","","","internet",3306,"/var/run/mysqld/mysqld.sock",0);
    if(handle==NULL)
        throw runtime_error("Unable to connec to database");
}

Mysql::~Mysql() {
    mysql_close(handle);
}

boost::shared_ptr<MYSQL_RES> Mysql::query(const std::string& query) {
    execute(query);
    boost::shared_ptr<MYSQL_RES> result(mysql_use_result(handle), mysql_free_result);
    if(result.get() == NULL)
        throw mysql_exception(handle, "mysql_use_result");
    return result;
}


std::string Mysql::prepare(std::string query, const params_t& params) {
    for(std::string::size_type i=0; i<query.size(); ++i){
        if(query[i]=='$') {
            std::string::size_type j;
            for(j = i+1; j<query.size() ; ++j) {
                if(isalnum(query[j]) || query[j]=='_')
                    continue;
                break;
            }
            std::string var = query.substr(i+1, j-(i+1));
            params_t::const_iterator val = params.find(var);
            if(val == params.end())
                throw std::runtime_error("Parameter ["+var+"] not found in attribs");
            char buf[250];
            mysql_real_escape_string(
                handle,
                buf,
                val->second.c_str(),
                val->second.size()
            );
            query.replace(i, j-i, buf);
            i = i + val->second.size();
            if(i>=query.size())
                break;
        }
    }
    return query;
}

File: store/jsp/src/store/Mysql.java
package store;


import javax.servlet.ServletException;
import java.sql.*;
import java.util.Map;
import java.util.Set;
import java.util.TreeMap;
import java.util.TreeSet;


public class Mysql {
    private Mysql() {
    }

    private static Map pool = new TreeMap();

    public static Connection get() throws ServletException {
        Integer thread_id = new Integer(Thread.currentThread().hashCode());
        Connection connection = (Connection) pool.get(thread_id);
        if(connection==null) {
            connection = createNew();
            pool.put(thread_id, connection);
        }
        return connection;
    }

    private static Connection createNew() throws ServletException {
        try {
            return DriverManager.getConnection("jdbc:mysql://localhost/internet");
        } catch (Throwable e) {
            e.printStackTrace();
            throw new ServletException(e);
        }
    }

    public synchronized void release(Connection conn) {
        try {
            conn.close();
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    public static void free(ResultSet result, Statement statement) {
        if (result != null) {
            try {
                result.close();
            } catch (SQLException ex) {  ex.printStackTrace(); }
            if (statement != null) {
                try {
                    statement.close();
                } catch (SQLException ex) {  ex.printStackTrace(); }
            }
        }
    }

    public static ResultSet query(Connection connection, String sql) throws SQLException {
        return connection.createStatement().executeQuery(sql);
    }

    public static String prepare(String query1, Map params) {
        StringBuffer query = new StringBuffer(query1);
        for(int i=0; i<query.length(); ++i){
            if(query.charAt(i)=='$') {
                int j;
                for(j = i+1; j<query.length() ; ++j) {
                    if(Character.isLetterOrDigit(query.charAt(j)) || query.charAt(j)=='_')
                        continue;
                    break;
                }
                String var = query.substring(i+1, j);
                final String val = (String) params.get(var);
                if(val == null)
                    throw new RuntimeException("Parameter ["+var+"] not found in attribs");
//                char buf[250];
//                mysql_real_escape_string(
//                    handle,
//                    buf,
//                    val->second.c_str(),
//                    val->second.size()
//                );
                query.replace(i, j, val);
                i = i + val.length();
                if(i>=query.length())
                    break;
            }
        }
        return query.toString();
    }

    public static boolean execute(Connection connection, String sql) throws SQLException {
        return connection.createStatement().execute(sql);
    }

}
